File: util/lang/ProcLang.java
package common.util.lang;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import common.CommonStatic;
import common.io.json.FieldOrder;
import common.io.json.JsonClass;
import common.io.json.JsonClass.RType;
import common.io.json.JsonClass.WType;
import common.io.json.JsonDecoder;
import common.io.json.JsonEncoder;
import common.io.json.JsonField;
import common.io.json.JsonField.IOType;
import common.pack.UserProfile;
import common.util.Data;
import common.util.Data.Proc;
import common.util.Data.Proc.IntType;
import common.util.lang.LocaleCenter.Binder;
import common.util.lang.LocaleCenter.DisplayItem;
import common.util.lang.LocaleCenter.ObjBinder;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Field;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

@JsonClass(read = RType.MANUAL, write = WType.CLASS, generator = "gen", serializer = "ser")
public class ProcLang {

	@JsonClass(read = RType.FILL)
	public static class ItemLang {
		@JsonField
		public String abbr_name, full_name, tooltip, format;

		private final String name;
		private final Class<?> cls;
		private final LinkedHashMap<String, DisplayItem> map = new LinkedHashMap<>();

		private ItemLang(String name, Class<?> cls) {
			this.cls = cls;
			this.name = name;
		}

		public Binder get(String proc) {
			try {
				return new ObjBinder(map.get(proc), name + "." + proc, this::getBinder);
			} catch (Exception e) {
				e.printStackTrace();
				return null;
			}
		}

		public String[] list() {
			String[] ans = new String[map.size()];
			int i = 0;
			for (Entry<String, DisplayItem> ent : map.entrySet())
				ans[i++] = ent.getKey();
			return ans;
		}

		@JsonField(tag = "class", io = IOType.R)
		public void readClass(JsonElement elem) {
			fill("", cls, elem == null ? null : elem.getAsJsonObject());
		}

		@JsonField(tag = "class", io = IOType.W)
		public JsonObject writeClass() {
			JsonObject ans = new JsonObject();
			fill(ans, "", cls);
			return ans;
		}

		private void fill(JsonObject ans, String pre, Class<?> c) {
			for (Field f : FieldOrder.getDeclaredFields(c)) {
				if (IntType.class.isAssignableFrom(f.getType()))
					fill(ans, f.getName() + ".", f.getType());
				else
					ans.add(pre + f.getName(), JsonEncoder.encode(map.get(pre + f.getName())));
			}
		}

		private void fill(String pre, Class<?> c, JsonObject obj) {
			for (Field f : FieldOrder.getDeclaredFields(c)) {
				if (IntType.class.isAssignableFrom(f.getType())) {
					fill(f.getName() + ".", f.getType(), obj);
				} else {
					JsonElement elem = obj == null ? null : obj.get(pre + f.getName());
					DisplayItem pf = elem == null ? new DisplayItem() : JsonDecoder.decode(elem, DisplayItem.class);
					map.put(pre + f.getName(), pf);
				}
			}
		}

		private Binder getBinder(String proc) {
			return ProcLang.get().get(name).get(proc);
		}

	}

	public static class ProcLangStore {

		private final ProcLang[] langs = new ProcLang[CommonStatic.Lang.LOC_CODE.length];

		private ProcLang getLang() {
			int lang = CommonStatic.getConfig().lang;
			if (langs[lang] == null)
				Data.err(ProcLang::read);
			return langs[lang];
		}

		private void setLang(ProcLang lang) {
			langs[CommonStatic.getConfig().lang] = lang;
		}

	}

	public static void clear() {
		UserProfile.setStatic("ProcLangStore", null);
	}

	public static ProcLang gen(JsonElement elem) throws Exception {
		ProcLang ans = new ProcLang();
		JsonObject obj = elem == null ? null : elem.getAsJsonObject();
		for (Field f : Proc.getDeclaredFields()) {
			String name = f.getName();
			ItemLang item = new ItemLang(name, f.getType());
			if (obj != null && obj.has(name))
				JsonDecoder.inject(obj.get(name), ItemLang.class, item);
			else
				item.readClass(null);
			ans.map.put(name, item);
		}
		return ans;
	}

	public static ProcLang get() {
		return store().getLang();
	}

	private static void read() throws Exception {
		InputStream f;

		if(CommonStatic.getConfig().lang == 2)
			f = CommonStatic.ctx.getLangFile("proc_kr.json");
		else if(CommonStatic.getConfig().lang == 3) {
			f = CommonStatic.ctx.getLangFile("proc_jp.json");
		} else
			f = CommonStatic.ctx.getLangFile("proc.json");

		JsonElement elem = JsonParser.parseReader(new InputStreamReader(f, StandardCharsets.UTF_8));
		f.close();
		ProcLang proc = JsonDecoder.decode(elem, ProcLang.class);
		store().setLang(proc);
	}

	private static ProcLangStore store() {
		return UserProfile.getStatic("ProcLangStore", ProcLangStore::new);
	}

	private final Map<String, ItemLang> map = new LinkedHashMap<>();

	private ProcLang() {
	}

	public ItemLang get(int i) {
		return get(Proc.getDeclaredFields()[i].getName());
	}

	public ItemLang get(String str) {
		return map.get(str);
	}

	public JsonObject ser() {
		JsonObject obj = new JsonObject();
		for (Field f : Proc.getDeclaredFields()) {
			String name = f.getName();
			obj.add(name, JsonEncoder.encode(map.get(name)));
		}
		return obj;
	}

}

File: util/lang/ProcLang.kt
package common.util.lang

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import common.CommonStatic
import common.io.json.JsonClass
import common.io.json.JsonClass.RType
import common.io.json.JsonClass.WType
import common.io.json.JsonDecoder
import common.io.json.JsonEncoder
import common.io.json.JsonField
import common.io.json.JsonField.IOType
import common.pack.Context
import common.pack.Context.ErrType
import common.pack.UserProfile
import common.util.Data
import common.util.Data.Proc
import common.util.lang.LocaleCenter.DisplayItem
import common.util.lang.LocaleCenter.ObjBinder
import common.util.lang.LocaleCenter.ObjBinder.BinderFunc
import common.util.lang.ProcLang
import java.io.File
import java.io.FileReader
import java.util.*
import java.util.function.Supplier

@JsonClass(read = RType.MANUAL, write = WType.CLASS, generator = "gen", serializer = "ser")
class ProcLang private constructor() {
    @JsonClass(read = RType.FILL)
    class ItemLang(private val name: String, private val cls: Class<*>) {
        @JsonField
        var abbr_name: String? = null

        @JsonField
        var full_name: String? = null

        @JsonField
        var tooltip: String? = null

        @JsonField
        var format: String? = null
        private val map: LinkedHashMap<String, DisplayItem> = LinkedHashMap<String, DisplayItem>()
        operator fun get(proc: String): LocaleCenter.Binder? {
            return try {
                ObjBinder(map[proc], "$name.$proc", BinderFunc { proc: String -> getBinder(proc) })
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }

        fun list(): Array<String?> {
            val ans = arrayOfNulls<String>(map.size)
            var i = 0
            for ((key) in map) ans[i++] = key
            return ans
        }

        @JsonField(tag = "class", io = IOType.R)
        fun readClass(elem: JsonElement?) {
            fill("", cls, if (elem == null) null else elem.asJsonObject)
        }

        @JsonField(tag = "class", io = IOType.W)
        fun writeClass(): JsonObject {
            val ans = JsonObject()
            fill(ans, "", cls)
            return ans
        }

        private fun fill(ans: JsonObject, pre: String, c: Class<*>) {
            for (f in c.declaredFields) {
                if (Data.Proc.IntType::class.java.isAssignableFrom(f.type)) fill(ans, f.name + ".", f.type) else ans.add(pre + f.name, JsonEncoder.Companion.encode(map[pre + f.name]))
            }
        }

        private fun fill(pre: String, c: Class<*>, obj: JsonObject?) {
            for (f in c.declaredFields) {
                if (Data.Proc.IntType::class.java.isAssignableFrom(f.type)) {
                    fill(f.name + ".", f.type, obj)
                } else {
                    val elem: JsonElement? = if (obj == null) null else obj.get(pre + f.name)
                    val pf: DisplayItem = if (elem == null) DisplayItem() else JsonDecoder.Companion.decode<DisplayItem>(elem, DisplayItem::class.java)
                    map[pre + f.name] = pf
                }
            }
        }

        private fun getBinder(proc: String): LocaleCenter.Binder? {
            return get()!![name]!![proc]
        }
    }

    class ProcLangStore {
        private val langs = arrayOfNulls<ProcLang>(CommonStatic.Lang.Companion.LOC_CODE.size)
        fun getLang(): ProcLang? {
            val lang: Int = CommonStatic.getConfig().lang
            if (langs[lang] == null) Data.Companion.err(Context.RunExc { read() })
            return langs[lang]
        }

        fun setLang(lang: ProcLang) {
            langs[CommonStatic.getConfig().lang] = lang
        }
    }

    private val map: MutableMap<String, ItemLang> = LinkedHashMap()
    operator fun get(i: Int): ItemLang {
        return get(Proc::class.java.declaredFields.get(i).name)
    }

    operator fun get(str: String): ItemLang? {
        return map[str]
    }

    fun ser(): JsonObject {
        val obj = JsonObject()
        for (f in Proc::class.java.declaredFields) {
            val name = f.name
            obj.add(name, JsonEncoder.Companion.encode(map[name]))
        }
        return obj
    }

    companion object {
        fun clear() {
            UserProfile.Companion.setStatic("ProcLangStore", null)
        }

        @Throws(Exception::class)
        fun gen(elem: JsonElement?): ProcLang {
            val ans = ProcLang()
            val obj: JsonObject? = if (elem == null) null else elem.asJsonObject
            for (f in Proc::class.java.declaredFields) {
                val name = f.name
                val item = ItemLang(name, f.type)
                if (obj != null && obj.has(name)) JsonDecoder.Companion.inject<ItemLang>(obj.get(name), ItemLang::class.java, item) else item.readClass(null)
                ans.map[name] = item
            }
            return ans
        }

        fun get(): ProcLang? {
            return store().getLang()
        }

        @Throws(Exception::class)
        private fun read() {
            val f: File = CommonStatic.ctx.getLangFile("proc.json")
            if (!f.exists()) CommonStatic.ctx.printErr(ErrType.FATAL, "cannot find proc.json")
            val elem: JsonElement = JsonParser.parseReader(FileReader(f))
            val proc: ProcLang = JsonDecoder.Companion.decode<ProcLang>(elem, ProcLang::class.java)
            store().setLang(proc)
        }

        private fun store(): ProcLangStore {
            return UserProfile.Companion.getStatic<ProcLangStore>("ProcLangStore", Supplier { ProcLangStore() })
        }
    }
}
